### PR TITLE
GC+ Link/Get physical address

### LINKED ISSUE #4476: GC+: Modify [Get/Link]PhysicalAddress
1. GetPhysicalAddress to return a validation token along with the address (or embedded as part of the address).
2. The token will be valid for a specified amount of time and for a single use.
3. lakeFS will need to track issued tokens/addresses, and delete them when tokens are expired/used
4. LinkPhysicalAddress to verify token valid before creating an entry.
    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
       the link API and avoid them being deleted by the GC process.
    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
       eventually be deleted by the GC job.

Related [issue](https://github.com/treeverse/lakeFS/issues/4438)

### PR DESCRIPTION
**Pull request by eden-ohana**
closes #4476

### CODE REVIEW

File: pkg/api/controller.go [hunk 1]
```diff
@@ -378,6 +384,14 @@ func (c *Controller) LinkPhysicalAddress(w http.ResponseWriter, r *http.Request,
 	writeTime := time.Now()
 	physicalAddress, addressType := normalizePhysicalAddress(repo.StorageNamespace, StringValue(body.Staging.PhysicalAddress))
 
+	// validate token
+	err = c.Catalog.GetAddressToken(ctx, repository, physicalAddress)
+	if err != nil {
+		writeError(w, r, http.StatusInternalServerError, err)
+		return
+	}
```
Comment (N-o-Z) [inline pkg/api/controller.go:392]: We should delete the token we are currently using, before we try create an entry

File: pkg/api/controller.go [hunk 2]
```diff
@@ -378,6 +384,14 @@ func (c *Controller) LinkPhysicalAddress(w http.ResponseWriter, r *http.Request,
 	writeTime := time.Now()
 	physicalAddress, addressType := normalizePhysicalAddress(repo.StorageNamespace, StringValue(body.Staging.PhysicalAddress))
 
+	// validate token
+	err = c.Catalog.GetAddressToken(ctx, repository, physicalAddress)
+	if err != nil {
+		writeError(w, r, http.StatusInternalServerError, err)
```
Comment (N-o-Z) [inline pkg/api/controller.go:390]: This should go through handleAPIError. I don't think internal server error is appropriate for trying to link with a token that doesn't exist

File: pkg/catalog/catalog.go [hunk 3]
```diff
@@ -1894,9 +1938,36 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
...
+		}
+
+		if newMark == nil {
+			// TODO(in progress) - fix hasData
+			newMark, hasData, err = c.writeTokensLocal(ctx, repository, uw, pw, mark, runID)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1968]: You can simplify the condition and also put it in the right logical order:

```suggestion
	if mark == nil || mark.BranchID != "" {
		newMark, hasData, err = c.writeUncommittedLocal(ctx, repository, uw, pw, mark, runID)
		if err != nil {
			return nil, err
		}
	}

	if newMark == nil {
		newMark, hasData, err = c.writeTokensLocal(ctx, repository, uw, pw, mark, runID)
		if err != nil {
			return nil, err
		}
	}
```

File: pkg/catalog/catalog.go [hunk 4]
```diff
@@ -1768,21 +1768,15 @@ type UncommittedParquetObject struct {
...
+func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, pw *writer.ParquetWriter, mark *GCUncommittedMark, runID string) (*GCUncommittedMark, bool, error) {
 	count := 0
 	itr, err := NewUncommittedIterator(ctx, c.Store, repository)
 	if err != nil {
 		return nil, false, err
 	}
 	defer itr.Close()
 
-	if mark != nil {
+	if mark != nil && mark.BranchID != "" {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1779]: Not needed - already checked outside

File: pkg/catalog/catalog.go [hunk 5]
```diff
@@ -1838,12 +1832,62 @@ func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *gravele
...
+func (c *Catalog) writeTokensLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, pw *writer.ParquetWriter, mark *GCUncommittedMark, runID string) (*GCUncommittedMark, bool, error) {
+	count := 0
+	itr, err := c.Store.ListAddressTokens(ctx, repository)
 	if err != nil {
 		return nil, false, err
 	}
+	defer itr.Close()
 
-	// Finished reading all staging area - no continuation token
+	if mark != nil && mark.BranchID == "" {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1849]: No need to check BranchID - already checked outside

File: pkg/catalog/catalog.go [hunk 6]
```diff
@@ -1838,12 +1832,62 @@ func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *gravele
...
+		itr.SeekGE(mark.Path.String())
+	}
+
+	for itr.Next() {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1853]: Consider reusing the iteration logic and just providing an iterator and a write function to it

File: pkg/catalog/catalog.go [hunk 7]
```diff
@@ -1894,9 +1938,36 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
...
+			return nil, err
+		}
+	} else {
+		newMark, hasData, err = c.writeUncommittedLocal(ctx, repository, uw, pw, mark, runID)
+		if err != nil {
+			return nil, err
+		}
+
+		if newMark == nil {
+			// TODO(in progress) - fix hasData
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1962]: You can pass hadData as reference to the writeTokensLocal and "modify" it to true if we've written token data

File: pkg/api/controller.go (hunk 1 above)
Comment (N-o-Z) [reply pkg/api/controller.go:392]: This will ensure we use the token only once

File: pkg/graveler/ref/manager.go [hunk 8]
```diff
@@ -7,6 +7,8 @@ import (
 	"strings"
 	"time"
 
+	"google.golang.org/protobuf/types/known/timestamppb"
+
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager.go:11]: Extra line

File: pkg/graveler/ref/manager.go [hunk 9]
```diff
@@ -27,6 +29,8 @@ const MaxBatchDelay = time.Millisecond * 3
 // commitIDStringLength string representation length of commit ID - based on hex representation of sha256
 const commitIDStringLength = 64
 
+const AddressTokenTime = 6 * time.Hour
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager.go:32]: Consider batching all the const declarations as below

File: pkg/graveler/ref/manager_test.go [hunk 10]
```diff
@@ -1025,3 +1025,69 @@ func TestManager_ListCommits(t *testing.T) {
 		i++
 	}
 }
+
+func TestManager_ListAddressTokens(t *testing.T) {
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager_test.go:1029]: We should add a test for expiry as well.

File: pkg/graveler/errors.go [hunk 11]
```diff
@@ -34,13 +34,15 @@ var (
 	ErrCommitNotFound               = fmt.Errorf("commit %w", ErrNotFound)
 	ErrCreateBranchNoCommit         = fmt.Errorf("can't create a branch without commit")
 	ErrRepositoryNotFound           = fmt.Errorf("repository %w", ErrNotFound)
+	ErrAddressTokenNotFound         = fmt.Errorf("address token %w", ErrNotFound)
```
Comment (N-o-Z) [inline pkg/graveler/errors.go:37]: Can we batch the errors together?

File: pkg/graveler/ref/manager.go [hunk 12]
```diff
@@ -561,3 +565,38 @@ func newCache(cfg *CacheConfig, def *CacheConfig) cache.Cache {
...
+	if err != nil {
+		if errors.Is(err, kv.ErrNotFound) {
+			err = graveler.ErrAddressTokenNotFound
+		}
+		return err
+	}
+	if data.ExpiredAt.AsTime().Before(time.Now()) {
+		return graveler.ErrAddressTokenExpired
+	}
+	return m.kvStore.DeleteMsg(ctx, graveler.RepoPartition(repository), path)
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager.go:597]: I think it's best to separate the Get and Delete functionalities. It can cause confusion and it doesn't really save us any operations

File: pkg/graveler/model.go [hunk 13]
```diff
@@ -66,6 +67,10 @@ func SettingsPath(key string) string {
 	return kv.FormatPath(settingsPrefix, key)
 }
 
+func AddressPath(key string) string {
+	return kv.FormatPath(addressesPrefix, key)
+}
```
Comment (N-o-Z) [inline pkg/graveler/model.go:72]:
```suggestion
func LinkedAddressPath(key string) string {
	return kv.FormatPath(addressesPrefix, key)
}
```

File: pkg/graveler/model.go [hunk 14]
```diff
@@ -15,6 +15,7 @@ const (
 	branchesPrefix         = "branches"
 	commitsPrefix          = "commits"
 	settingsPrefix         = "settings"
+	addressesPrefix        = "addresses"
```
Comment (N-o-Z) [inline pkg/graveler/model.go:18]:
```suggestion
	linkAddressesPrefix        = "link-addresses"
```

File: pkg/graveler/graveler.proto [hunk 15]
```diff
@@ -75,4 +75,10 @@ message StagedEntryData {
   bytes key = 1;
   bytes identity = 2;
   bytes data = 3;
+}
+
+// message data model to track address tokens
+message AddressData {
```
Comment (N-o-Z) [inline pkg/graveler/graveler.proto:81]:
```suggestion
message LinkAddressData {
```

File: pkg/graveler/graveler.proto [hunk 16]
```diff
@@ -75,4 +75,10 @@ message StagedEntryData {
...
+  string address = 1;
+  google.protobuf.Timestamp expired_at = 2;
```
Comment (N-o-Z) [inline pkg/graveler/graveler.proto:83]: Newline at end of file

File: pkg/graveler/ref/address_token_iterator.go [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+		err:           nil,
+	}, nil
+}
+
+func (i *AddressTokenIterator) Next() bool {
+	if i.Err() != nil || i.closed {
+		return false
+	}
+	if !i.it.Next() {
+		i.value = nil
```
Comment (N-o-Z) [inline pkg/graveler/ref/address_token_iterator.go:44]: I think you can put this line at the top of the function

File: pkg/graveler/ref/address_token_iterator.go [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+		return false
+	}
+	e := i.it.Entry()
+	if e == nil {
+		i.err = graveler.ErrReadingFromStore
```
Comment (N-o-Z) [inline pkg/graveler/ref/address_token_iterator.go:49]:
```suggestion
		i.err = e
```

File: pkg/graveler/ref/address_token_iterator.go [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+		return false
+	}
+	token, ok := e.Value.(*graveler.AddressData)
```
Comment (N-o-Z) [inline pkg/graveler/ref/address_token_iterator.go:52]: This can't really happen so I recommend removing it altogether

File: pkg/graveler/ref/address_token_iterator.go [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+	i.value = token
+	return true
+}
+
+func (i *AddressTokenIterator) SeekGE(address string) {
+	if i.Err() != nil {
+		return
+	}
+	i.Close()
+	it, err := kv.NewPrimaryIterator(i.ctx, i.store, (&graveler.AddressData{}).ProtoReflect().Type(),
```
Comment (N-o-Z) [inline pkg/graveler/ref/address_token_iterator.go:66]:
```suggestion
	it, i.err := kv.NewPrimaryIterator(i.ctx, i.store, (&graveler.AddressData{}).ProtoReflect().Type(),
```

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks for the great changes and all the efforts put into this!
I've started by reviewing without tests as there are some open issues

File: pkg/api/controller.go (hunk 1 above)
Comment (eden-ohana) [reply pkg/api/controller.go:392]: GetAddressToken also deletes the token

File: pkg/api/controller.go (hunk 2 above)
Comment (eden-ohana) [reply pkg/api/controller.go:390]: done

File: pkg/catalog/catalog.go (hunk 3 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1968]: done

File: pkg/catalog/catalog.go (hunk 4 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1779]: done

File: pkg/catalog/catalog.go (hunk 5 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1849]: not done :)

File: pkg/graveler/ref/manager.go (hunk 8 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:11]: done

File: pkg/graveler/errors.go (hunk 11 above)
Comment (eden-ohana) [reply pkg/graveler/errors.go:37]: what do you mean?

File: pkg/graveler/model.go (hunk 13 above)
Comment (eden-ohana) [reply pkg/graveler/model.go:72]: done

File: pkg/graveler/graveler.proto (hunk 15 above)
Comment (eden-ohana) [reply pkg/graveler/graveler.proto:81]: done

File: pkg/graveler/graveler.proto (hunk 16 above)
Comment (eden-ohana) [reply pkg/graveler/graveler.proto:83]: done

File: pkg/graveler/errors.go (hunk 11 above)
Comment (N-o-Z) [reply pkg/graveler/errors.go:37]: Put all the addressToken errors together one after the other

Comment (eden-ohana) [reply pkg/graveler/errors.go]: done

File: pkg/catalog/catalog.go (hunk 7 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go]: done

File: pkg/graveler/ref/manager.go (hunk 12 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:576]: done

File: pkg/graveler/ref/address_token_iterator.go (hunk 17 above)
Comment (eden-ohana) [reply pkg/graveler/ref/address_token_iterator.go:44]: I'm not sure why

File: pkg/graveler/ref/address_token_iterator.go (hunk 18 above)
Comment (eden-ohana) [reply pkg/graveler/ref/address_token_iterator.go:49]: e is not an error

File: pkg/graveler/ref/address_token_iterator.go (hunk 19 above)
Comment (eden-ohana) [reply pkg/graveler/ref/address_token_iterator.go]: why it can't happen? if we saved under this prefix not address data

File: pkg/graveler/ref/address_token_iterator.go (hunk 20 above)
Comment (eden-ohana) [reply pkg/graveler/ref/address_token_iterator.go]: done

File: clients/hadoopfs/src/main/java/io/lakefs/LinkOnCloseOutputStream.java [hunk 21]
```diff
@@ -18,6 +18,7 @@ class LinkOnCloseOutputStream extends OutputStream {
     private final URI physicalUri;
     private final MetadataClient metadataClient;
     private final OutputStream out;
+    private boolean isLinked;
```
Comment (N-o-Z) [inline clients/hadoopfs/src/main/java/io/lakefs/LinkOnCloseOutputStream.java:21]: Can these changes be in a separate PR?

File: pkg/upload/path_provider.go [hunk 22]
```diff
@@ -101,6 +102,15 @@ func (i *PathPartitionProvider) NewPath() string {
 	return path.Join(i.cfg.Prefix, part, name)
 }
 
+func (i *PathPartitionProvider) ResolvePathTime(addressPath string) (time.Time, error) {
+	_, address := path.Split(addressPath)
+	id, err := xid.FromString(address)
```
Comment (N-o-Z) [inline pkg/upload/path_provider.go:107]:
```suggestion
func (i *PathPartitionProvider) ResolvePathTime(address string) (time.Time, error) {
	_, name := path.Split(address)
	id, err := xid.FromString(name)
```

File: pkg/upload/path_provider.go [hunk 23]
```diff
@@ -11,6 +11,7 @@ import (
 // PathProvider captures the requirements from PathPartitionProvider implementation to return a new path
 type PathProvider interface {
 	NewPath() string
+	ResolvePathTime(addressPath string) (time.Time, error)
```
Comment (N-o-Z) [inline pkg/upload/path_provider.go:14]:
```suggestion
	ResolvePathTime(address string) (time.Time, error)
```

File: pkg/graveler/ref/address_token_iterator.go (hunk 18 above)
Comment (N-o-Z) [reply pkg/graveler/ref/address_token_iterator.go:49]: Right, so if i.it.Next() returns true and i.it.Entry == nil - there's a bug in the iterator `it` not sure we need to even check it in this context

File: pkg/graveler/ref/manager.go [hunk 24]
```diff
@@ -547,11 +546,11 @@ func newCache(cfg CacheConfig) cache.Cache {
 }
 
 func (m *Manager) SetAddressToken(ctx context.Context, repository *graveler.RepositoryRecord, token string) error {
-	a := &graveler.AddressData{
+	a := &graveler.LinkAddressData{
 		Address:   token,
 		ExpiredAt: timestamppb.New(time.Now().Add(AddressTokenTime)),
```
Comment (N-o-Z) [inline pkg/graveler/ref/manager.go:551]: Why do we still need the ExpiredAt if it's embedded in the address?

File: pkg/graveler/ref/manager.go (hunk 12 above)
Comment (N-o-Z) [reply pkg/graveler/ref/manager.go:580]: I meant that we also should not delete the address as part of the GetAddressToken functionality but call it explicitly where needed

File: pkg/catalog/catalog.go [hunk 25]
```diff
@@ -2020,6 +2005,14 @@ func (c *Catalog) GetAddressToken(ctx context.Context, repository, token string)
 	return c.Store.GetAddressToken(ctx, repo, token)
 }
 
+func (c *Catalog) DeleteExpiredAddressTokens(ctx context.Context, repository string) error {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:2008]: As far as I understand this is not used in the scope of this PR and should move to:
https://github.com/treeverse/lakeFS/pull/4936

File: pkg/api/controller.go (hunk 1 above)
Comment (N-o-Z) [reply pkg/api/controller.go:404]: I don't think that this is a good behavior. We should call deleteToken explicitly wherever required

File: pkg/catalog/catalog.go [hunk 26]
```diff
@@ -1951,24 +1943,17 @@ func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string,
...
+	if mark == nil || mark.BranchID != "" {
+		newMark, hasData, err = c.writeUncommittedLocal(ctx, repository, uw, pw, mark, runID)
 		if err != nil {
 			return nil, err
 		}
-	} else {
-		newMark, hasData, err = c.writeUncommittedLocal(ctx, repository, uw, pw, mark, runID)
+	}
+	if newMark == nil {
+		newMark, hasData, err = c.writeTokensLocal(ctx, repository, uw, pw, mark, runID, hasData)
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1953]: You are assigning hasData and passing it to the function, you can just pass it by reference and save on the returned values

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Thanks @eden-ohana looks like we're getting close.
Some more comments

File: clients/hadoopfs/src/main/java/io/lakefs/LinkOnCloseOutputStream.java (hunk 21 above)
Comment (eden-ohana) [reply clients/hadoopfs/src/main/java/io/lakefs/LinkOnCloseOutputStream.java]: done

File: pkg/upload/path_provider.go (hunk 22 above)
Comment (eden-ohana) [reply pkg/upload/path_provider.go]: done

File: pkg/upload/path_provider.go (hunk 23 above)
Comment (eden-ohana) [reply pkg/upload/path_provider.go]: done

File: pkg/graveler/ref/manager.go (hunk 24 above)
Comment (eden-ohana) [reply pkg/graveler/ref/manager.go:551]: deleted

File: pkg/catalog/catalog.go (hunk 25 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:2008]: #4936 is for the delete job. To test it I need to use Set and Get that's why I need it in this PR

File: pkg/catalog/catalog.go (hunk 26 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1953]: If I'm passing it by reference I need to assign it in addition to the return value. I'm not sure it's better

File: pkg/catalog/catalog.go (hunk 25 above)
Comment (N-o-Z) [reply pkg/catalog/catalog.go:2008]: I think there's a misunderstanding all the DeleteExpiredAddressTokens functions (catalog and graveler) are not really used in this PR and should move to the #4936 as this is the right context for them (correct me if I'm wrong)

File: pkg/catalog/catalog.go [hunk 27]
```diff
@@ -1842,13 +1832,59 @@ func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *gravele
...
 		return nil, false, itr.Err()
 	}
-	err = pw.WriteStop()
+
+	// Finished reading all staging area - no continuation token
+	hasData := count > 0
+	return nil, hasData, nil
+}
+
+func (c *Catalog) writeTokensLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, pw *writer.ParquetWriter, mark *GCUncommittedMark, runID string, hasData bool) (*GCUncommittedMark, bool, error) {
```
Comment (itaiad200) [inline pkg/catalog/catalog.go:1841]: Why do we need this? Isn't the grace period in the client sufficient?

**Review by itaiad200 — CHANGES_REQUESTED** [review]
Partial review - Thanks!

File: pkg/api/controller.go [hunk 28]
```diff
@@ -377,6 +384,24 @@ func (c *Controller) LinkPhysicalAddress(w http.ResponseWriter, r *http.Request,
...
+		c.handleAPIError(ctx, w, r, err)
+		return
+	}
+	addressTokenExpiryTime := addressTokenCreationTime.Add(ref.AddressTokenTime)
+	if addressTokenExpiryTime.Before(time.Now()) {
+		c.handleAPIError(ctx, w, r, graveler.ErrAddressTokenExpired)
+		return
+	}
+
+	err = c.Catalog.GetAddressToken(ctx, repository, physicalAddress)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/api/controller.go:399]: returns nil if the token is valid (exists and not expired) and deletes it

File: pkg/api/controller.go [hunk 29]
```diff
@@ -1847,6 +1872,12 @@ func (c *Controller) handleAPIErrorCallback(ctx context.Context, w http.Response
 	case errors.Is(err, graveler.ErrTooManyTries):
 		cb(w, r, http.StatusLocked, "Too many attempts, try again later")
 
+	case errors.Is(err, graveler.ErrAddressTokenExpired):
+		cb(w, r, http.StatusGone, "address token is expired")
+
+	case errors.Is(err, graveler.ErrAddressTokenNotFound):
+		cb(w, r, http.StatusNotFound, "address token not found")
+
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/api/controller.go:1880]: done

File: pkg/catalog/catalog.go (hunk 26 above)
Comment (N-o-Z) [reply pkg/catalog/catalog.go:1953]: Lets sync

File: pkg/catalog/catalog.go (hunk 27 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go:1841]: deleted :(((((((

File: pkg/graveler/graveler.go [hunk 30]
```diff
@@ -717,6 +741,21 @@ type RefManager interface {
 	// GCCommitIterator TODO (niro): Remove when DB implementation is deleted
 	// GCCommitIterator temporary WA to support both DB and KV GC CommitIterator
 	GCCommitIterator(ctx context.Context, repository *RepositoryRecord) (CommitIterator, error)
+
+	// GetAddressToken verifies the given address token
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/graveler/graveler.go:742]: chaged to VerifyAddressToken

File: pkg/api/controller.go [hunk 31]
```diff
@@ -329,6 +330,12 @@ func (c *Controller) GetPhysicalAddress(w http.ResponseWriter, r *http.Request,
 		return
 	}
 
+	err = c.Catalog.SetAddressToken(ctx, repository, address)
+	if err != nil {
+		writeError(w, r, http.StatusInternalServerError, err)
```
[the remark this answers is not in the document]
Comment (eden-ohana) [reply pkg/api/controller.go:335]: done

File: pkg/graveler/ref/address_token_iterator.go (hunk 18 above)
Comment (itaiad200) [reply pkg/graveler/ref/address_token_iterator.go:49]: You can switch from `e` to `ent`/`entry` to avoid the confusion with errors

File: pkg/graveler/graveler.go [hunk 32]
```diff
@@ -717,6 +738,18 @@ type RefManager interface {
...
+	VerifyAddressToken(ctx context.Context, repository *RepositoryRecord, token string) error
+
+	// SetAddressToken creates address token
+	SetAddressToken(ctx context.Context, repository *RepositoryRecord, token string) error
+
+	// ListAddressTokens lists address tokens on a repository
+	ListAddressTokens(ctx context.Context, repository *RepositoryRecord) (AddressTokenIterator, error)
+
+	// IsTokenExpired returns nil if the token is valid and not expired
+	IsTokenExpired(token *LinkAddressData) (bool, error)
```
Comment (itaiad200) [inline pkg/graveler/graveler.go:752]: I think this belongs in a dedicated manager. Not a blocker though

File: pkg/graveler/ref/manager.go [hunk 33]
```diff
@@ -27,6 +30,18 @@ const MaxBatchDelay = time.Millisecond * 3
...
+const (
+	DefaultRepositoryCacheSize   = 1000
+	DefaultRepositoryCacheExpiry = 5 * time.Second
+	DefaultRepositoryCacheJitter = DefaultRepositoryCacheExpiry / 2
+
+	DefaultCommitCacheSize   = 1000
+	DefaultCommitCacheExpiry = 30 * time.Second
+	DefaultCommitCacheJitter = DefaultRepositoryCacheExpiry / 2
+
+	AddressTokenTime = 6 * time.Hour
```
Comment (itaiad200) [inline pkg/graveler/ref/manager.go:42]: Please document this value

File: pkg/api/controller.go [hunk 34]
```diff
@@ -377,6 +384,24 @@ func (c *Controller) LinkPhysicalAddress(w http.ResponseWriter, r *http.Request,
...
+	addressTokenCreationTime, err := c.PathProvider.ResolvePathTime(physicalAddress)
+	if err != nil {
+		writeError(w, r, http.StatusInternalServerError, err)
+		return
+	}
+	addressTokenExpiryTime := addressTokenCreationTime.Add(ref.AddressTokenTime)
+	if addressTokenExpiryTime.Before(time.Now()) {
+		c.handleAPIError(ctx, w, r, graveler.ErrAddressTokenExpired)
+		return
+	}
```
Comment (itaiad200) [inline pkg/api/controller.go:397]: It feels to me like this belongs in the `VerifyAddressToken`

**Review by itaiad200 — APPROVED** [review]
Approved with some comments, thanks!

File: pkg/catalog/catalog.go [hunk 35]
```diff
@@ -1772,7 +1772,13 @@ type UncommittedParquetObject struct {
 	CreationDate    int64  `parquet:"name=creation_date, type=INT64, convertedtype=INT_64"`
 }
 
-func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, pw *writer.ParquetWriter, mark *GCUncommittedMark, runID string) (*GCUncommittedMark, bool, error) {
+func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *graveler.RepositoryRecord, w *UncommittedWriter, mark *GCUncommittedMark, runID string) (*GCUncommittedMark, bool, error) {
```
Comment (N-o-Z) [inline pkg/catalog/catalog.go:1775]: Since we don't write the linked addresses anymore - I think we can revert the changes in writeUncommittedLocal

**Review by N-o-Z — CHANGES_REQUESTED** [review]
@eden-ohana  lets revert the changes in writeUncommittedLocal

File: pkg/api/controller.go (hunk 34 above)
Comment (eden-ohana) [reply pkg/api/controller.go]: removed. thanks

File: pkg/catalog/catalog.go (hunk 35 above)
Comment (eden-ohana) [reply pkg/catalog/catalog.go]: done

**Review by N-o-Z — APPROVED** [review]
LGTM! Thanks for the very hard work! 💪🏽
